### ISSUE TITLE
2.6 mouse issues

### ISSUE THREAD

**Issue by chinagaofeng** [Genymobile/scrcpy#5162]
as 2.6 said:

**By default, a simple click triggers the shortcuts, and Shift+click forwards the click to the device.**

I hate this default setting! It makes very inconvenience for control than 2.5. I have to roll-back.

my cli is just "scrcpy -Sw", means I use SDK mouse

**Comment by rom1v** [comment]
> I hate this default setting! It makes very inconvenience for control than 2.5. I have to roll-back.

If you just click (with a secondary button), it is the same behavior as v2.5… and earlier versions.

What changes is only when you <kbd>Shift</kbd>+click, there are new bindings.

**Comment by chinagaofeng** [comment]
> > I hate this default setting! It makes very inconvenience for control than 2.5. I have to roll-back.
> 
> If you just click (with a secondary button), it is the same behavior as v2.5… and earlier versions.
> 
> What changes is only when you Shift+click, there are new bindings.

not exactly same, for example: in 2.5-, holding left button and dragging, works same as middle button scrolling; but in 2.6, holding-dragging works but mid-scrolling not works, or mid-scrolling works while holding-dragging not works, can't determine why

**Comment by vferrenq** [comment]
I encounter also a problem with 2.6 (Ubuntu 22.04). A simple left click acts as a long press. I have also to rollback to 2.5 and everything works as expected.

**Comment by rom1v** [comment]
OK, that's a bug then. Not caused by the binding changes directly, but probably by either 1fee79bf50b124223523eb51c437c1267c2724a or 6baea57987a1867f2157f5c1001e77ca3cb1c6c5.

I will bisect and investigate (hopefully tonight or tomorrow).

**Comment by vidoardes** [comment]
I am also experiencing this issue, is there an easy way to downgrade on Fedora? I tried `sudo dnf downgrade scrcpy` but it appears previous versions of the package has been removed

**Comment by rom1v** [comment]
> but in 2.6, holding-dragging works but mid-scrolling not works, or mid-scrolling works while holding-dragging not works

OK, so now I'm on my computer, I tried to reproduce, but I fail. For example, in the settings, I can either scroll or click & drag.

Is it "random"? Or 100% reproducible?

Could you please give more information:
 - what device model?
 - in which app does it happen?

**Comment by rom1v** [comment]
Oh, I understand, in Firefox, I can scroll using middle-scroll, but not by click&drag (because now it behaves like a mouse rather than a finger). This is definitely due to 6808288823239b0f3a76f9be377e4de82e91b35a.

So I will need to restore a pointer, but that will cause other problems (for example, mouse hover needs a mouse: #5067). Will need to find something which works for all cases.

**Comment by rom1v** [comment]
Oh, using a "finger" tool type in all cases seems to be sufficient:

```diff
diff --git a/server/src/main/java/com/genymobile/scrcpy/control/Controller.java b/server/src/main/java/com/genymobile/scrcpy/control/Controller.java
index 85425113a..a5cda74cf 100644
--- a/server/src/main/java/com/genymobile/scrcpy/control/Controller.java
+++ b/server/src/main/java/com/genymobile/scrcpy/control/Controller.java
@@ -278,14 +278,16 @@ public class Controller implements AsyncProcessor {
         pointer.setPressure(pressure);
 
         int source;
+
+        // Always behave like a finger (to scroll a page via Click+drag)
+        pointerProperties[pointerIndex].toolType = MotionEvent.TOOL_TYPE_FINGER;
+
         if (pointerId == POINTER_ID_MOUSE) {
             // real mouse event
-            pointerProperties[pointerIndex].toolType = MotionEvent.TOOL_TYPE_MOUSE;
             source = InputDevice.SOURCE_MOUSE;
             pointer.setUp(buttons == 0);
         } else {
             // POINTER_ID_GENERIC_FINGER, POINTER_ID_VIRTUAL_FINGER or real touch from device
-            pointerProperties[pointerIndex].toolType = MotionEvent.TOOL_TYPE_FINGER;
             source = InputDevice.SOURCE_TOUCHSCREEN;
             // Buttons must not be set for touch events
             buttons = 0;
```

Please test by replacing this file in the scrcpy v2.6 release folder:

- [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5162/1/scrcpy-server) <sub>`SHA-256: 9ceb957ade7f2df9c59e8b8af8856239365687a05e4624c39bee2decff9f5e0`</sub>

**Comment by rom1v** [comment]
@vidoardes you can download the scrcpy-server above and test with:

```bash
export SCRCPY_SERVER_PATH=/xxx/scrcpy-server
scrcpy
```

**Comment by vidoardes** [comment]
> @vidoardes you can download the scrcpy-server above and test with:
> 
> ```shell
> export SCRCPY_SERVER_PATH=/xxx/scrcpy-server  # adapt the path
> scrcpy
> ```

I can confirm it fixes scrolling in Firefox, but not in all apps. Not sure what the difference is, but one example is a game call Gold & Goblins which doesn't now work (but did in 2.5)

**Comment by rom1v** [comment]
OK, thank you very much. Now I can reproduce and debug easily :+1:

The the trick in https://github.com/Genymobile/scrcpy/issues/5162#issuecomment-2265735461 is not correct: it works only if the input is a `SOURCE_TOUCHSCREEN`. So back to the previous comment (https://github.com/Genymobile/scrcpy/issues/5162#issuecomment-2265716809). Will need more thinking to make everything work with a single solution.

**Comment by rom1v** [comment]
Please test with this change (89288c9e94718183512b3e7f33aa496ed7a9af9c):

- [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5162/2/scrcpy-server) <sub>`SHA-256: 6a4a5f2f14a511ab64e8717e0c0c1fb173a57cad4f53cc6ad8e19aba6715de8`</sub>

**Comment by aoy2024** [comment]
Sorry, I don't understand code. I can't read it

**Comment by aoy2024** [comment]
> Please test with this change ([89288c9](https://github.com/Genymobile/scrcpy/commit/89288c9e94718183512b3e7f33aa496ed7a9af9c)):
> 
> * [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5162/2/scrcpy-server) `SHA-256: 6a4a5f2f14a511ab64e8717e0c0c1fb173a57cad4f53cc6ad8e19aba6715de8`

Sorry, I don't understand code. I can't read it

**Comment by aoy2024** [comment]
> Please test with this change ([89288c9](https://github.com/Genymobile/scrcpy/commit/89288c9e94718183512b3e7f33aa496ed7a9af9c)):
> 
> * [`scrcpy-server`](https://tmp.rom1v.com/scrcpy/5162/2/scrcpy-server) `SHA-256: 6a4a5f2f14a511ab64e8717e0c0c1fb173a57cad4f53cc6ad8e19aba6715de8`

After replacing the file, I can click normally and the problem has been solved. Thank you very much

**Comment by rom1v** [comment]
Fixed in scrcpy 2.6.1.
